hoomd/mpcd/RejectionVirtualParticleFiller.h

```diff
@@ -0,0 +1,175 @@
+
+template<class Geometry>
+class PYBIND11_EXPORT RejectionVirtualParticleFiller : public mpcd::VirtualParticleFiller
+    {
+    public:
+        //! Constructor
+        /*!
+         * \param sysdata - MPCD system data
+         * \param density - number density of MPCD solvent
+         * \param type - particle fill type
+         * \param T - Temperature
+         * \param seed - seed for PRNG
+         * \param geom - confinement geometry
+         */
+        RejectionVirtualParticleFiller(std::shared_ptr<mpcd::SystemData> sysdata,
+                                       Scalar density,
+                                       unsigned int type,
+                                       std::shared_ptr<::Variant> T,
+                                       unsigned int seed,
+                                       std::shared_ptr<const Geometry> geom)
+        : mpcd::VirtualParticleFiller(sysdata, density, type, T, seed), m_geom(geom)
+        {
+        m_exec_conf->msg->notice(5) << "constructing MPCD RejectionVirtualParticleFiller" << std::endl;
+        }
+
+        //! Destructor
+        ~RejectionVirtualParticleFiller()
+        {
+        m_exec_conf->msg->notice(5) << "Destroying MPCD RejectionVirtualParticleFiller" << std::endl;
+        }
+
+        //! Get the streaming geometry
+        std::shared_ptr<const Geometry> getGeometry() const
+            {
+            return m_geom;
+            }
+
+        //! Set the streaming geometry
+        void setGeometry(std::shared_ptr<const Geometry> geom)
+            {
+            m_geom = geom;
+            }
+
+        //! Fill the particles outside the confinement
+        void fill(unsigned int timestep);
+
+    protected:
+        std::shared_ptr<const Geometry> m_geom;
+        GPUArray<Scalar4> m_tmp_pos;
+        GPUArray<Scalar4> m_tmp_velTag;
+    };
+
+
+template<class Geometry>
+void RejectionVirtualParticleFiller<Geometry>::fill(unsigned int timestep)
+    {
+    // Number of particles that we need to draw (constant)
+    const BoxDim& box = m_pdata->getBox();
+    const Scalar3 lo = box.getLo();
+    const Scalar3 hi = box.getHi();
+    const unsigned int NVirtMax = round(m_density*box.getVolume());
+
+    // Step 1: Create temporary GPUArrays to draw Particles locally using the worst case estimate for number
+    //         number of particles.
+    if (NVirtMax > m_tmp_pos.getNumElements())
+        {
+        GPUArray<Scalar4> tmp_pos(NVirtMax, m_exec_conf);
+        GPUArray<Scalar4> tmp_velTag(NVirtMax, m_exec_conf);
+        m_tmp_pos.swap(tmp_pos);
+        m_tmp_velTag.swap(tmp_velTag);
+        }
+    ArrayHandle<Scalar4> h_tmp_pos(m_tmp_pos, access_location::host, access_mode::overwrite);
+    ArrayHandle<Scalar4> h_tmp_velTag(m_tmp_velTag, access_location::host, access_mode::overwrite);
+
+    // Step 2: Draw the particles and assign velocities simultaneously by using temporary memory
+    unsigned int pidx = 0;
+    unsigned int tag = computeFirstTag(&pidx);
```

Comment:
I think this is going to give weird behavior. I suggest we use a fictitious tag here, then compact the tag range in the second loop...
```suggestion
    unsigned int pidx = 0;
    const unsigned int first_tag = computeFirstTag(NVirtMax);
```
(See comment below about using value rather than pointer)

Reply:
I was thinking of not making it a const variable so that we can reuse it when we want to compute the actual tags. What do you think?

Reply:
Yes, that is perfectly fine! I probably did not notice this variable would be reassigned when I marked it `const` originally.